math/firep.h

```diff
@@ -0,0 +1,123 @@
+ * \author  Roy Zhao; edited by Michael Lesnick.
+ * \date    March 2017; edited September 2017.
+ */
+
+#ifndef __FIRep_H__
+#define __FIRep_H__
+
+//forward declarations
+class IndexMatrix;
+class MapMatrix;
+class MapMatrix_Perm;
+
+#include "bifiltration_data.h"
+
+#include <set>
+#include <string>
+#include <vector>
+
+//used to build the hash tables for simplices in hom_dim-1 and hom_dim
+struct VectorHash {
+    std::size_t operator()(Simplex const* const& v) const
+    {
+        return boost::hash_range(v->begin(), v->end());
+    }
+};
+
+struct deref_equal_fn {
+    bool operator()(Simplex* const& lhs, Simplex* const& rhs) const
+    {
+        return *lhs == *rhs;
+    }
+};
+
+typedef std::unordered_map<Simplex* const, unsigned, VectorHash, deref_equal_fn> SimplexHashLow;
+typedef std::unordered_map<Simplex* const, std::vector<MidHighSimplexData>::iterator, VectorHash, deref_equal_fn> SimplexHashMid;
+//no need for a hash table in the high dimension
+
+class FIRep {
+
+public:
+    FIRep(BifiltrationData& bd, int v); //constructor; requires verbosity parameter
+
+    FIRep(BifiltrationData& bd, int t, int s, int r, std::vector<std::vector<unsigned>>& d2, std::vector<std::vector<unsigned>>& d1,
+        const std::vector<unsigned> x_values, const std::vector<unsigned> y_values, int v); //constructor
+
+    ~FIRep(); //destructor
+
+    //returns a matrix of boundary information for dim to dim-1
+    MapMatrix* get_low_boundary_mx();
+
+    //returns a matrix of boundary information for dim+1 to dim
+    MapMatrix* get_high_boundary_mx();
+
+    //returns a boundary matrix for hom_dim-simplices with columns in a specified order -- for vineyard-update algorithm
+    MapMatrix_Perm* get_boundary_mx(std::vector<int>& coface_order, unsigned num_simplices);
+
+    //returns a boundary matrix for (hom_dim+1)-simplices with columns and rows a specified orders -- for vineyard-update algorithm
+    MapMatrix_Perm* get_boundary_mx(std::vector<int>& face_order, unsigned num_faces, std::vector<int>& coface_order, unsigned num_cofaces);
+
+    //TODO: shouldn't the following two return pointers to consts?
+
+    //returns a matrix of column indexes to accompany MapMatrices
+    IndexMatrix* get_low_index_mx();
+
+    //returns a matrix of column indexes to accompany MapMatrices
+    IndexMatrix* get_high_index_mx();
+
+    unsigned num_x_grades(); //returns the number of unique x-coordinates of the multi-grades
+    unsigned num_y_grades(); //returns the number of unique y-coordinates of the multi-grades
+
+    void print(); //Print the matrices and appearance grades
+
+    const unsigned verbosity; //controls display of output, for debugging
+
+private:
+    unsigned x_grades; //the number of x-grades that exist in this bifiltration
+    unsigned y_grades; //the number of y-grades that exist in this bifiltration
+    MapMatrix* boundary_mx_0; //boundary matrix from dim to dim-1
```

Comment:
Why pointers? Can they ever be null? Should they be references instead? Or unique_ptr?